### PR TITLE
[FLINK-30983][runtime] Support configured ssl algorithms for external REST SSL

### PR DESCRIPTION
**Pull request by TanYuxin-tyx**
## What is the purpose of the change

*Support configured ssl algorithms for external REST SSL*

## Brief change log

*(for example:)*
  - *Support configured ssl algorithms for external REST SSL*

## Verifying this change

Add a test in SSLUtilsTest.
  - *Add test in SSLUtilsTest*

## Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency): (yes / **no**)
  - The public API, i.e., is any changed class annotated with `@Public(Evolving)`: (yes / **no**)
  - The serializers: (yes / **no** / don't know)
  - The runtime per-record code paths (performance sensitive): (yes / **no** / don't know)
  - Anything that affects deployment or recovery: JobManager (and its components), Checkpointing, Kubernetes/Yarn, ZooKeeper: (yes / **no** / don't know)
  - The S3 file system connector: (yes / **no** / don't know)

## Documentation

  - Does this pull request introduce a new feature? (yes / **no**)
  - If yes, how is the feature documented? (**not applicable** / docs / JavaDocs / not documented)

### CODE REVIEW

File: flink-runtime/src/test/java/org/apache/flink/runtime/net/SSLUtilsTest.java [hunk 1]
```diff
@@ -362,26 +388,25 @@ public void testSetSSLVersionAndCipherSuitesForSSLServerSocket() throws Exceptio
...
             String[] algorithms = sslSocket.getEnabledCipherSuites();
 
-            assertEquals(1, protocols.length);
-            assertEquals("TLSv1.1", protocols[0]);
-            assertEquals(2, algorithms.length);
-            assertThat(
-                    algorithms,
-                    arrayContainingInAnyOrder(
-                            "TLS_RSA_WITH_AES_128_CBC_SHA", "TLS_RSA_WITH_AES_128_CBC_SHA256"));
+            assertThat(protocols.length).isEqualTo(1);
```
Comment (reswqa) [inline flink-runtime/src/test/java/org/apache/flink/runtime/net/SSLUtilsTest.java:397]:
```suggestion
            assertThat(protocols).hasSize(1);
```

Comment (reswqa) [inline flink-runtime/src/test/java/org/apache/flink/runtime/net/SSLUtilsTest.java:397]:
```suggestion
            assertThat(protocols).hasSize(1);
```

File: flink-runtime/src/test/java/org/apache/flink/runtime/net/SSLUtilsTest.java [hunk 2]
```diff
@@ -362,26 +388,25 @@ public void testSetSSLVersionAndCipherSuitesForSSLServerSocket() throws Exceptio
 
         try (ServerSocket socket =
                 SSLUtils.createSSLServerSocketFactory(serverConfig).createServerSocket(0)) {
-            assertTrue(socket instanceof SSLServerSocket);
+            assertThat(socket instanceof SSLServerSocket).isTrue();
```
Comment (reswqa) [inline flink-runtime/src/test/java/org/apache/flink/runtime/net/SSLUtilsTest.java:391]:
```suggestion
           assertThat(socket).isInstanceOf(SSLServerSocket.class);
```

File: flink-runtime/src/test/java/org/apache/flink/runtime/net/SSLUtilsTest.java [hunk 3]
```diff
@@ -362,26 +388,25 @@ public void testSetSSLVersionAndCipherSuitesForSSLServerSocket() throws Exceptio
...
-            assertEquals(1, protocols.length);
-            assertEquals("TLSv1.1", protocols[0]);
-            assertEquals(2, algorithms.length);
-            assertThat(
-                    algorithms,
-                    arrayContainingInAnyOrder(
-                            "TLS_RSA_WITH_AES_128_CBC_SHA", "TLS_RSA_WITH_AES_128_CBC_SHA256"));
+            assertThat(protocols.length).isEqualTo(1);
+            assertThat(protocols[0]).isEqualTo("TLSv1.1");
+            assertThat(algorithms.length).isEqualTo(2);
```
Comment (reswqa) [inline flink-runtime/src/test/java/org/apache/flink/runtime/net/SSLUtilsTest.java:399]:
```suggestion
            assertThat(algorithms).hasSize(2);
```

File: flink-runtime/src/test/java/org/apache/flink/runtime/net/SSLUtilsTest.java [hunk 4]
```diff
@@ -171,21 +169,38 @@ public void testRESTClientSSLWrongPassword() throws Exception {
...
+    public void testRESTSSLConfigCipherAlgorithms(String sslProvider) throws Exception {
+        String testSSLAlgorithms = "test_algorithm1,test_algorithm2";
+        Configuration config = createRestSslConfigWithTrustStore(sslProvider);
+        config.setBoolean(SecurityOptions.SSL_REST_ENABLED, true);
+        config.setString(SecurityOptions.SSL_ALGORITHMS.key(), testSSLAlgorithms);
+        JdkSslContext nettySSLContext =
+                (JdkSslContext)
+                        SSLUtils.createRestNettySSLContext(config, true, ClientAuth.NONE, JDK);
+        List<String> cipherSuites = checkNotNull(nettySSLContext).cipherSuites();
+        assertThat(cipherSuites.size()).isEqualTo(2);
```
Comment (reswqa) [inline flink-runtime/src/test/java/org/apache/flink/runtime/net/SSLUtilsTest.java:183]:
```suggestion
        assertThat(cipherSuites).hasSize(2);
```

File: flink-runtime/src/test/java/org/apache/flink/runtime/net/SSLUtilsTest.java [hunk 5]
```diff
@@ -171,21 +169,38 @@ public void testRESTClientSSLWrongPassword() throws Exception {
...
+        assertThat(cipherSuites.toArray()).containsExactlyInAnyOrder(testSSLAlgorithms.split(","));
```
Comment (reswqa) [inline flink-runtime/src/test/java/org/apache/flink/runtime/net/SSLUtilsTest.java:184]:
```suggestion
        assertThat(cipherSuites).containsExactlyInAnyOrder(testSSLAlgorithms.split(","));
```

File: flink-runtime/src/test/java/org/apache/flink/runtime/net/SSLUtilsTest.java [hunk 6]
```diff
@@ -408,20 +433,19 @@ public void testCreateSSLEngineFactory() throws Exception {
         final SslHandler sslHandler =
                 serverSSLHandlerFactory.createNettySSLHandler(UnpooledByteBufAllocator.DEFAULT);
 
-        assertEquals(expectedSslProtocols.length, sslHandler.engine().getEnabledProtocols().length);
-        assertThat(
-                sslHandler.engine().getEnabledProtocols(),
-                arrayContainingInAnyOrder(expectedSslProtocols));
+        assertThat(sslHandler.engine().getEnabledProtocols().length)
+                .isEqualTo(expectedSslProtocols.length);
```
Comment (reswqa) [inline flink-runtime/src/test/java/org/apache/flink/runtime/net/SSLUtilsTest.java:437]:
```suggestion
        assertThat(sslHandler.engine().getEnabledProtocols())
                .hasSameSizeAs(expectedSslProtocols);
```

File: flink-runtime/src/test/java/org/apache/flink/runtime/net/SSLUtilsTest.java [hunk 7]
```diff
@@ -408,20 +433,19 @@ public void testCreateSSLEngineFactory() throws Exception {
...
+        assertThat(sslHandler.engine().getEnabledProtocols()).contains(expectedSslProtocols);
 
-        assertEquals(sslAlgorithms.length, sslHandler.engine().getEnabledCipherSuites().length);
-        assertThat(
-                sslHandler.engine().getEnabledCipherSuites(),
-                arrayContainingInAnyOrder(sslAlgorithms));
+        assertThat(sslHandler.engine().getEnabledCipherSuites().length)
+                .isEqualTo(sslAlgorithms.length);
```
Comment (reswqa) [inline flink-runtime/src/test/java/org/apache/flink/runtime/net/SSLUtilsTest.java:441]:
```suggestion
        assertThat(sslHandler.engine().getEnabledCipherSuites())
                .hasSameSizeAs(sslAlgorithms);
```

File: flink-runtime/src/test/java/org/apache/flink/runtime/net/SSLUtilsTest.java [hunk 8]
```diff
@@ -432,21 +456,21 @@ public void testInvalidFingerprintParsing() throws Exception {
             SSLUtils.createInternalServerSSLEngineFactory(config);
             fail("expected exception");
```
Comment (reswqa) [inline flink-runtime/src/test/java/org/apache/flink/runtime/net/SSLUtilsTest.java:457]: We'd better avoid using `fail` in the junit test now. Please also check the other place has same problem.

**Review by reswqa — COMMENTED** [review]
Thanks @TanYuxin-tyx, this changes overall look good to me. I only left some comments about tests, PTAL~
BTW, I think we also need to introduce an IT case for this, maybe `RestServerSSLAuthITCase` is a good place.

**Comment by TanYuxin-tyx** [comment]
> Thanks @TanYuxin-tyx, this changes overall look good to me. I only left some comments about tests, PTAL~ BTW, I think we also need to introduce an IT case for this, maybe `RestServerSSLAuthITCase` is a good place.
Thanks for reviewing the change. @reswqa . The IT case RestServerEndpointITCase#setup has covered the modified method.

**Comment by TanYuxin-tyx** [comment]
Thanks for reviewing the change @reswqa . The IT case RestServerEndpointITCase#setup has covered the modified method.

File: flink-runtime/src/test/java/org/apache/flink/runtime/net/SSLUtilsTest.java (hunk 8 above)
Comment (TanYuxin-tyx) [reply flink-runtime/src/test/java/org/apache/flink/runtime/net/SSLUtilsTest.java]: Fixed, removed the fail method.

File: flink-runtime/src/test/java/org/apache/flink/runtime/net/SSLUtilsTest.java (hunk 7 above)
Comment (TanYuxin-tyx) [reply flink-runtime/src/test/java/org/apache/flink/runtime/net/SSLUtilsTest.java]: Fixed.

**Review by reswqa — APPROVED** [review]
Thanks @TanYuxin-tyx, LGTM! Would you mind backport this to release-1.17 branch?

**Comment by TanYuxin-tyx** [comment]
> Thanks @TanYuxin-tyx, LGTM! Would you mind backport this to release-1.17 branch?

OK, Thanks @reswqa , I have opened a new PR for 1.17. https://github.com/apache/flink/pull/22097

**Pull request closed without merging**